src/lib/Type.hs

```diff
@@ -117,20 +118,17 @@ instance HasType Atom where
       zipWithM_ (|:) params paramTys
       let paramTysRemaining = drop (length params) paramTys
       return $ foldr (-->) TyKind paramTysRemaining
+    LabeledRow row -> checkLabeledRow row $> LabeledRowKind
     Record items -> do
       types <- mapM typeCheck items
-      return $ RecordTy types
-    RecordTy items -> do
-      mapM_ (|: TyKind) items
-      return TyKind
-    Variant vtys@(LabeledItems types) label i value -> do
-      value |: ((types M.! label) !! i)
+      return $ RecordTy (NoExt types)
+    RecordTy row -> checkLabeledRow row $> TyKind
+    Variant vtys@(Ext (LabeledItems types) _) label i value -> do
+      value |: ((types M.! label) NE.!! i)
```

Comment:
Wouldn't that crash the compiler if you actually had a mis-typed variant? The `!` functions are usually partial

Reply:
I guess you're right. I'll add a check first.